test/core/dvectorTests.cpp

```diff
@@ -0,0 +1,920 @@
+
+        it -= 2;                   // 1
+        EXPECT_EQ(it->first, -30); // 3
+
+        EXPECT_EQ((it - 1)->first, -40); // 3
+
+        EXPECT_EQ(it - (it + 211), -211);
+
+        EXPECT_EQ(it[-2].first, -50); // 3
+
+        // Comparision
+
+        EXPECT_TRUE(it == it);
+        EXPECT_FALSE(it == (it + 10));
+
+        EXPECT_TRUE(it != (it + 10));
+        EXPECT_FALSE(it != it);
+
+        EXPECT_TRUE(it < (it + 10));
+        EXPECT_FALSE(it < (it - 10));
+        EXPECT_FALSE(it < it);
+
+        EXPECT_TRUE(it > (it - 10));
+        EXPECT_FALSE(it > (it + 10));
+        EXPECT_FALSE(it > it);
+
+        EXPECT_TRUE(it <= (it + 10));
+        EXPECT_FALSE(it <= (it - 10));
+        EXPECT_TRUE(it <= it);
+
+        EXPECT_TRUE(it >= (it - 10));
+        EXPECT_FALSE(it >= (it + 10));
+        EXPECT_TRUE(it >= it);
+    }
+
+    TEST(DVector, IteratorConstness) {
+        DVector<int> v;
+
+        static_assert(!std::is_const_v<std::remove_reference_t<decltype(*v.begin())>>);
+        static_assert(std::is_const_v<std::remove_reference_t<decltype(*v.cbegin())>>);
+
+        static_assert(!std::is_const_v<std::remove_reference_t<decltype(*v.end())>>);
+        static_assert(std::is_const_v<std::remove_reference_t<decltype(*v.cend())>>);
+
+        // Test for implicit conversion from non const to const
+        EXPECT_EQ(v.begin() - v.cbegin(), 0);
+        EXPECT_EQ(v.begin(), v.cbegin());
+        EXPECT_EQ(v.cbegin(), v.begin());
+
+        EXPECT_EQ(v.rbegin(), v.crbegin());
+        EXPECT_EQ(v.crbegin(), v.rbegin());
+        EXPECT_EQ(v.begin(), v.crend().base() + 1);
+    }
+
+    // Capacity
+
+    TEST(DVector, FrontBackSize) {
+        DVector<int> v;
+        EXPECT_EQ(v.size_front(), 0);
+        EXPECT_EQ(v.size_back(), 0);
+
+        v.push_front(-10);
+        v.push_front(-20);
+
+        v.push_back(0);
+        v.push_back(10);
+        v.push_back(20);
+        v.push_back(30);
+
+        EXPECT_EQ(v.size_front(), 2);
+        EXPECT_EQ(v.size_back(), 4);
+    }
+    TEST(DVector, ShrinkCapacityFitSize) {
+        DVector<int> v;
+        v.reserve(100);
+
+        v.push_back(0);
+        v.push_back(0);
+
+        v.shrink_to_fit();
+
+        EXPECT_EQ(v.capacity(), 3);
+    }
+
+    // Modifiers
+
+    TEST(DVector, Clear) {
+        DVector<int> v;
+
+        v.reserve(3);
+
+        v.push_front(0);
+        v.push_back(1);
+        v.push_back(2);
+
+        v.clear();
+
+        v.push_front(9);
+        v.push_back(10);
+        EXPECT_EQ(v.capacity(), 3);
+        EXPECT_EQ(v[-1], 9);
+        EXPECT_EQ(v[0], 10);
+    }
+
+    TEST(DVector, ClearCallDestructor) {
+        TestClass::ResetCounter();
+
+        DVector<TestClass> v;
+        v.reserve(3);
+
+        v.emplace_front();
+        v.emplace_back();
+        v.emplace_back();
+
+        v.clear();
+        EXPECT_EQ(TestClass::destructorCalls, 3);
+    }
+
+    TEST(DVector, EmplaceMoveFront) {
```

Comment:
Remove emplace, in the future add emplace_move_front and emplace_move_back
Current emplace offers no way to modify at index -1